### PR TITLE
Fix over increment test flake in guarddog test.

### LINKED ISSUE #1141: Flake in guarddog_impl_test
Triggered in https://github.com/lyft/envoy/pull/1134, see https://travis-ci.org/lyft/envoy/jobs/244689841:

[----------] 2 tests from GuardDogMissTest
[ RUN      ] GuardDogMissTest.MissTest
external/envoy/test/server/guarddog_impl_test.cc:164: Failure
      Expected: 1UL
      Which is: 1
To be equal to: stats_store_.counter("server.watchdog_miss").value()
      Which is: 2
[  FAILED  ] GuardDogMissTest.MissTest (647 ms)

### PR DESCRIPTION
**Pull request by dnoe**
Whether due to spurious wakeups, actual timeouts, etc, the GuardDog loop
might run several times after a thread ceases responding and the counter
(miss or megamiss) would get incremented multiple times even though the
thread only stopped responding once. This caused a test flake and was
also a departure from the old watchdog stat behavior where the
resolution of a starvation event resulted in a single increment.

With this change the GuardDog will only count one miss and one megamiss
for each starvation event. If the watchdog is touched this resets and we
will increment the stats again if there is another period of
nonresponsiveness.

Fixes #1141

### CODE REVIEW

File: test/server/guarddog_impl_test.cc [hunk 1]
```diff
@@ -189,6 +189,48 @@ TEST_F(GuardDogMissTest, MegaMissTest) {
   unpet_dog = nullptr;
 }
 
+TEST_F(GuardDogMissTest, MissCountTest) {
+  // This tests a flake discovered in the MissTest where real timeout or
+  // spurious condition_variable wakeup causes the counter to get incremented
+  // more than it should be.
+  ON_CALL(time_source_, currentTime())
+      .WillByDefault(testing::Invoke([&]() {
```
Comment (htuch) [inline test/server/guarddog_impl_test.cc]: I prefer explicitly enumerating variables here, too much is implicitly captured by `&`.

File: test/server/guarddog_impl_test.cc [hunk 2]
```diff
@@ -189,6 +189,48 @@ TEST_F(GuardDogMissTest, MegaMissTest) {
...
+        return std::chrono::steady_clock::time_point(std::chrono::milliseconds(mock_time_));
+      }));
+  GuardDogImpl gd(stats_store_, config_miss_, time_source_);
+  auto sometimes_pet_dog = gd.createWatchDog(0);
+  for (unsigned long i = 0; i < 3; i++) {
```
Comment (htuch) [inline test/server/guarddog_impl_test.cc]: Why the looping?

File: test/server/guarddog_impl_test.cc (hunk 1 above)
Comment (dnoe) [inline test/server/guarddog_impl_test.cc]: Sure, I can make this change (and the other spots where this idiom is repeated).

File: test/server/guarddog_impl_test.cc (hunk 2 above)
Comment (dnoe) [inline test/server/guarddog_impl_test.cc]: The test needs to verify that after we touch the watchdog the behavior resets back to baseline.  So it does this by executing the test steps twice, with the last step being the touch that should reset the behavior.

The third iteration is unnecessary, I'll remove it.

**Comment by dnoe** [comment]
Yes regarding manual verification, and the test's use of repeated forceCheckForTest simulates exactly what happens if there is an early or spurious wakeup.

**Pull request merged**